test/expr/unary/keypath/keypath.swift

```diff
@@ -570,6 +571,7 @@ struct VisibilityTesting {
 struct VisibilityTesting2 {
   func inFilePrivateContext() {
     var xRef = \VisibilityTesting.x
+    // expected-warning@-1 {{forming a writable keypath to property 'x' that is read-only in this context is deprecated and will be removed in a future release}}
```

Comment:
How are you supposed to fix this one to be read-only, though?

Reply:
You can make it read-only with an explicit type, but in retrospect this warning should really be tracking the keypaths that would be formed as read-only in {{-swift-version 5}} that are used in key path applications that write.

I'll take a look at updating the PR to do that instead.

Reply:
In Swift 5 mode, this infers as a read-only `KeyPath`, right?

Reply:
Yes, that’s correct.

Reply:
Why not diagnose both?  You'll probably need to build a cache of incorrectly-inferred declrefs to catch stuff like this

```swift
class Foo {
  private(set) var foo: String = ""

  init() {}
}

let f = Foo();
let kp = \Foo.foo // forming a writable keypath to property...
f[keyPath: kp] = "asdf" // write to writeable key path that is read-only in this context is deprecated...
```

Reply:
I'm unhappy with giving people a warning on code that doesn't do anything wrong.

```swift
let f = Foo()
let kp = \Foo.foo // happens to be writable...
print(f[keyPath: kp]) // ...but I only use it to read
```

Reply:
@codafi Detecting the write in your example is what I meant when I said "...this warning should really be tracking the keypaths that would be formed as read-only in -swift-version 5 that are used in key path applications that write.".

As Jordan points out, we don't want to warn about things that we happen to infer today as writable but are never used that way.

Anything we do in Sema is going to be necessarily approximate because doing this properly requires data flow analysis.

I'm not sure how much value we'll actually get from doing something approximate, either. I don't want to end up with any false positives, and the only way to do that is to clear out the cache of values we believe hold writable keypaths that would have been inferred as readable any time we exit a control structure.

This may ultimately not be worth doing.